fdbrpc/Locality.h

```diff
@@ -180,6 +180,22 @@ struct LocalityData {
 	Optional<Standalone<StringRef>> dcId() const { return get(keyDcId); }
 	Optional<Standalone<StringRef>> dataHallId() const { return get(keyDataHallId); }
 
+	bool isValidProcesId() const {
+		return processId().present() && processId().get().size() > 0;
+	}
+	bool isValidZoneId() const {
+		return zoneId().present() && zoneId().get().size() > 0;
+	}
+	bool isValidMachineId() const {
+		return machineId().present() && machineId().get().size() > 0;
+	}
+	bool isValidDcId() const {
+		return dcId().present() && dcId().get().size() > 0;
+	}
+	bool isValidDataHallId() const {
+		return dataHallId().present() && dataHallId().get().size() > 0;
```

Comment:
If I'm understanding it correctly, it looks like the policy accepts empty values (though not missing values). I'm not sure what the right thing to do is, but it feels like the two locations should at least be in agreement.

Reply:
I suppose if it works right now for an empty data hall (or other locality field), then we shouldn't break it in a patch release at the very least. I can test it to see what happens.

Reply:
I think I should use a knob to guard this feature for safe.

Reply:
I think it may be reasonable to fail for missing locality fields, but my concern is about empty (i.e. size == 0). I think if empty localities are allowed now, then we should continue allowing them unconditionally.

Reply:
It seemed to work for me when running the same sort of test but with an empty data hall instead of a missing one. I think my proposal then would be something like:

```suggestion
		return dataHallId().present();
```

Reply:
OK. Now I agree that we should only check if it is set.

As to zoneID, I got your concern that the change may break existing clusters that do not set zoneID. I think I can fix it.